samd/ports/SAMD21_XPLAINED_PRO: Update the deploy instructions and use the external flash for the file system.

robert-hh opened the pull request:
### Summary

The SAMD21 Xplained Pro board is delivered without a bootloader, making it impossible to run MicroPython. This PR tells, how to install a .uf2 bootloader and provides samples of bootloaders, which can be used for the board. In addition it
adapts and fixes MicroPython to use the on-board 1MByte flash for the file system. Commits in this PR:

- Update the deploy instructions
- Proved two bootloaders as Intel Hex files. One of the is a generic SAMD21 bootloader, that should work on any SAMD21 board. The other has the Volume labels and USB identifiers changed to reflect the Xplained Pro board, but is besides that identical to the generic bootloader.
- Use the 1 MByte flash for the file system.

### Testing

- Building the firmware. 
- Updating the bootloader following the steps of the instructions.
- Updating the Firmware using all methods to activate the bootloader.
- Creating the file system at the 1MB flash (done automatically at the first boot)
- Creating files , storing data and deleting files with a stress test.

robert-hh committed:
samd/boards/SAMD21_XPLAINED_PRO: Use the SPI flash for the file system.

The initial settings did not support it.  The change required to add a
dedicated handling of the Adesto 1MByte flash of the XPLAINED PRO board,
which does not support the sfdp feature.

Fixes the ID check of the Adesto/Renesas 1MByte flash.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/boards/SAMD21_XPLAINED_PRO: Add specific deploy instructions.

Add instructions to install a bootloader to the board.  The SAMD21 XPLAINED
PRO board is shipped without a bootloader, which therefore has to be
installed once before it can be used with MicroPython.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/mboot: Provide a UF2 bootloader for SAMD21 Xplained Pro.

A bootloader labelled for the SAMD21 XPLAINED PRO board.  The only
difference to a generic bootloader are the names and ID of the USB port and
the label of the drive that is opened.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented:
Would it be better to instead change the linker script for this board so that the firmware is placed at the beginning of flash, so that a bootloader is not needed at all?

For a dev board like this one, I would assume that someone using it would perhaps want to use MicroPython on it as-is, without a bootloader.  After all, you need to use the debug port to program the bootloader initially.  So why not just program MicroPython using the debug port?

dpgeorge commented:
> That's possible, but then this board would behave different than all other SAMD ports, and it would need the .hex version of the firmware. At the moment there are only .uf2 versions available for the SAMD port.

OK, that's a good point.  So it would be easier to make this board behave like all other samd boards.

In that case it would be good to provide a bootloader.  But, I don't think it's appropriate to check it into this repo.  Instead can we just host it at micropython.org/resources/firmware?  I can copy it there and we can link to it in the deploy.md file.

robert-hh commented:
> Instead can we just host it at micropython.org/resources/firmware? I

That's good. I was wondering if there is a place to host support files likes this one. Should I keep the file in the PR?

robert-hh commented:
@dpgeorge The file deploy.md is updated to point to micropython.org/resources/firmware. And since it's a .hex files already, the step of converting a .bin file to Intel hex is not needed any more.

robert-hh committed:
samd/samd_flash: Make flash read/write methods access self parameters.

Use `self` (the first argument) instead of the global `samd_flash_obj` when
accessing the `flash_base` parameter.  This allows there to be multiple
flash objects for various types of filesystem.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented on ports/samd/boards/SAMD21_XPLAINED_PRO/deploy_xplained_pro.md:
```diff
@@ -0,0 +1,25 @@
+following procedure has been found to work and be simple:
+
+1. Get the bootloader file bootloader-xplained-pro-v3.16.0-15-gaa52b22.hex
+from https://micropython.org/resources/firmware.
```
I think it would be better if this URL was the full URL of the .hex file, like:
```
Get the bootloader from https://micropython.org/resources/firmware/bootloader-xplained-pro-v3.16.0-15-gaa52b22.hex
```

dpgeorge commented:
@robert-hh I have merged this and the .hex file for the bootloader is copied across to the micropython.org URL.

Actually, I think it would be good to add a README.md to the new `ports/samd/mboot/` directory which describes how that .hex file is generated (which repository it comes from), and also giving the license of that file (which would probably just be the license of the source code it was generated from).  Would you mind making a PR for that README?

robert-hh commented:
Sure. I will do that,
